### PR TITLE
caddytls: Add option to configure certificate lifetime

### PR DESCRIPTION
**Pull request by clauverjat**
Hello,

This PR allows users to specify the certificate validity period to ask the CA for. This PR is the result of the discussion on the corresponding issue #5985.

I took my initial implementation of the feature and updated it slightly so that it is compatible with the changes made by @mholt to CertMagic (I took the current master branch).
Unfortunately though, the resulting implementation does not work. This is because the certmagic "NewACMEIssuer"
function is overriding the NotAfter field that I set in caddy's makeIssuerTemplate(). Here is the problematic code :  https://github.com/caddyserver/certmagic/blob/master/acmeissuer.go#L193-L198 

I think this is likely a "bug" in CertMagic NewACMEIssuer function, since the behavior on NotBefore/NotAfter fields clashes with how the function deals with the other fields. I think you meant to write the following instead  : 
```
if template.NotBefore == 0 {
	template.NotBefore = DefaultACME.NotBefore
}
if template.NotAfter == 0 {
	template.NotAfter = DefaultACME.NotAfter
}
```

Let me know what you think,

Best regards
Corentin Lauverjat

### CODE REVIEW

**Comment by mholt** [comment]
Oops!! You're so right. I have pushed a fix here: https://github.com/caddyserver/certmagic/commit/855d4670a49d42499a373436852216fcd4e4047f

**Comment by clauverjat** [comment]
I just pushed a new commit with this update

File: modules/caddytls/acmeissuer.go [hunk 1]
```diff
@@ -88,6 +88,16 @@ type ACMEIssuer struct {
...
 	PreferredChains *ChainPreference `json:"preferred_chains,omitempty"`
 
+	// The validity period to ask the CA to issue a certificate for.
+	// Default: 0 (don't ask a custom lifetime to the CA)
+	// This value is used to compute the "notAfter" field of the ACME order,
+	// therefore the system must have a reasonably synchronized clock.
+	// Important : Let's Encrypt and ZeroSSL don't allow custom
+	// validity period and will refuse to issue a certicate if this is set.
+	// For CAs that support it, there are often limits
+	// on the allowed validity periods. Please refer to your CA documentation.
```
Comment (mholt) [inline modules/caddytls/acmeissuer.go:98]:
```suggestion
	// The validity period to ask the CA to issue a certificate for.
	// Default: 0 (CA chooses lifetime).
	// This value is used to compute the "notAfter" field of the ACME order;
	// therefore the system must have a reasonably synchronized clock.
	// NOTE: Not all CAs support this. Check with your CA's ACME
	// documentation to see if this is allowed and what values may
	// be used.
```

File: modules/caddytls/acmeissuer.go [hunk 2]
```diff
@@ -349,6 +360,17 @@ func (iss *ACMEIssuer) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
...
+		case "lifetime":
+			var lifetimeStr string
+			if !d.AllArgs(&lifetimeStr) {
+				return d.ArgErr()
+			}
+			lifetime, err := caddy.ParseDuration(lifetimeStr)
+			if err != nil {
+				return d.Errf("invalid lifetime %s: %v", lifetimeStr, err)
+			}
+			// TODO: Add check that valid lifetime must be >= 0
```
Comment (mholt) [inline modules/caddytls/acmeissuer.go:372]:
```suggestion
			if lifetime < 0 {
				return d.Errf("lifetime must be > 0: %s", lifetime)
			}
```

**Review by mholt — CHANGES_REQUESTED** [review]
Excellent, thank you!

Just a couple nits; also, can you run `go mod tidy` to make sure that go.sum file is cleaned up if needed?

File: modules/caddytls/acmeissuer.go (hunk 2 above)
Comment (clauverjat) [reply modules/caddytls/acmeissuer.go]: Just applied the suggested change with a small twist : the lifetime can be 0

File: modules/caddytls/acmeissuer.go [hunk 3]
```diff
@@ -88,6 +88,15 @@ type ACMEIssuer struct {
 	// will be selected.
 	PreferredChains *ChainPreference `json:"preferred_chains,omitempty"`
 
+	// The validity period to ask the CA to issue a certificate for.
+	// Default: 0 (CA chooses lifetime).
+	// This value is used to compute the "notAfter" field of the ACME order;
+	// therefore the system must have a reasonably synchronized clock.
+	// NOTE: Not all CAs support this. Check with your CA's ACME
+	// documentation to see if this is allowed and what values may
+	// be used.
```
Comment (mholt) [inline modules/caddytls/acmeissuer.go:97]:
```suggestion
	// be used. EXPERIMENTAL: Subject to change.
```

**Comment by mholt** [comment]
That `gci` failure on that line has got to be a mistake...

File: modules/caddytls/acmeissuer.go (hunk 3 above)
Comment (clauverjat) [reply modules/caddytls/acmeissuer.go]: I just did it. Sorry for the delay, I am quite busy at this moment with other work. I looked into using "Allow edits from maintainers" to give you access. But unfortunately it seems that this option is still not supported by GH for repo owned by org (which is the case here) : https://github.com/orgs/community/discussions/5634 :disappointed:

Comment (mholt) [reply modules/caddytls/acmeissuer.go]: Ohhh gotcha, no problem. Thanks for the patch! I just don't want to overpromise :)

**Comment by mholt** [comment]
Thanks for the contribution, we will try it out! If there are any issues with it, would you be able/willing to help with it? :blush:

**Pull request merged**